### PR TITLE
gh-102213: Optimize the performance of `__getattr__`

### PR DESCRIPTION
**Pull request by sunmy2019**
This is the logically corrected version of GH-102248.

GH-102248 caused several issues and had been reverted. This PR (I think) correctly implements it.

Tests passed on Ubuntu 22.04 (WSL) gcc-11 with `--pydebug`.

I also do the perf tests in the original PR, (`--enable-optimizations` with my AMD 5900X)
Before,
```
Python version of 3.12.0a7+ (heads/gh-102248-corrected:af53046995, Apr 24 2023, 22:42:22) [GCC 11.3.0]
Call __getattr__ spend time: 0.2576024532318115
```
After,
```
Python version of 3.12.0a7+ (heads/gh-102248-corrected:ac81fd6c7e, Apr 24 2023, 22:39:39) [GCC 11.3.0]
Call __getattr__ spend time: 0.04635906219482422
```

cc: @Fidget-Spinner

* Issue: gh-102213

### CODE REVIEW

File: Objects/typeobject.c [hunk 1]
```diff
@@ -8257,17 +8257,22 @@ _Py_slot_tp_getattr_hook(PyObject *self, PyObject *name)
     if (getattribute == NULL ||
         (Py_IS_TYPE(getattribute, &PyWrapperDescr_Type) &&
          ((PyWrapperDescrObject *)getattribute)->d_wrapped ==
-         (void *)PyObject_GenericGetAttr))
-        res = PyObject_GenericGetAttr(self, name);
-    else {
+             (void *)PyObject_GenericGetAttr)) {
+        res = _PyObject_GenericGetAttrWithDict(self, name, NULL, 1);
+        /* if no error has occureed, then it must be suppressed by us */
```
Comment (Eclips4) [inline Objects/typeobject.c]:
```suggestion
        /* if no error has occured, then it must be suppressed by us */
```

File: Objects/typeobject.c [hunk 2]
```diff
@@ -8257,17 +8257,22 @@ _Py_slot_tp_getattr_hook(PyObject *self, PyObject *name)
     if (getattribute == NULL ||
         (Py_IS_TYPE(getattribute, &PyWrapperDescr_Type) &&
          ((PyWrapperDescrObject *)getattribute)->d_wrapped ==
-         (void *)PyObject_GenericGetAttr))
-        res = PyObject_GenericGetAttr(self, name);
-    else {
+             (void *)PyObject_GenericGetAttr)) {
+        res = _PyObject_GenericGetAttrWithDict(self, name, NULL, 1);
+        /* if no error has occured, then it must be suppressed by us */
```
Comment (sunmy2019) [inline Objects/typeobject.c]:
```suggestion
        /* if no error has occurred, then it must be suppressed by us */
```

**Comment by Fidget-Spinner** [comment]
I will review and hopefully merge this on Friday when I have more time. Thanks.

**Comment by bluetech** [comment]
#103551 has a simple test case for the issue with the original PR, might be useful to add as a unit test for this PR.

**Comment by sunmy2019** [comment]
> #103551 has a simple test case for the issue with the original PR, might be useful to add as a unit test for this PR.

`@property` raises and `__getattr__` not called  --> added in #103358
`@property` calls `__getattr__`, and `__getattr__` raises --> added in #103336

The test of #103551 should be covered by #103358. But a little duplication is okay. I will add it later.

File: Objects/typeobject.c [hunk 3]
```diff
@@ -8296,17 +8296,22 @@ _Py_slot_tp_getattr_hook(PyObject *self, PyObject *name)
...
          ((PyWrapperDescrObject *)getattribute)->d_wrapped ==
-         (void *)PyObject_GenericGetAttr))
-        res = PyObject_GenericGetAttr(self, name);
-    else {
+             (void *)PyObject_GenericGetAttr)) {
+        res = _PyObject_GenericGetAttrWithDict(self, name, NULL, 1);
+        /* if no error has occurred, then it must be suppressed by us */
+        if (res == NULL && !PyErr_Occurred()) {
+            res = call_attribute(self, getattr, name);
+        }
```
Comment (Fidget-Spinner) [inline Objects/typeobject.c]: I don't understand this comment. Could you explain please?

Comment (sunmy2019) [reply Objects/typeobject.c]: With `_PyObject_GenericGetAttrWithDict(self, name, NULL, /*suppress=*/1);`,

`_PyObject_GenericGetAttrWithDict` will catch all `AttributeError`, and do not raise `AttributeError` at the end.

Other kinds of exceptions are untouched.

If no exception occurred with `res == NULL`, we already catch that in `_PyObject_GenericGetAttrWithDict`. This is the only case we should call the `call_attribute(self, getattr, name);`

Comment (sunmy2019) [reply Objects/typeobject.c]: I may change this comment to
```suggestion
        /* if res == NULL with no exception set, then it must be an
           AttributeError suppressed by us. */
        if (res == NULL && !PyErr_Occurred()) {
            res = call_attribute(self, getattr, name);
        }
```

Comment (Fidget-Spinner) [reply Objects/typeobject.c]: Ah yes that's better. Now it makes sense. thanks.

**Comment by sunmy2019** [comment]
Does anyone have any idea why this fails?
https://buildbot.python.org/all/#/builders/738/builds/3112

**Comment by Fidget-Spinner** [comment]
> Got a timeout here: https://buildbot.python.org/all/#/builders/403/builds/1066

This specific test seems flaky. I would ignore it.

**Comment by sunmy2019** [comment]
@Fidget-Spinner How is it going on? We are running out of time. New features will soon be frozen in May. 7th.
https://peps.python.org/pep-0693/

**Comment by Fidget-Spinner** [comment]
> @Fidget-Spinner How is it going on? We are running out of time. New features will soon be frozen in May. 7th. https://peps.python.org/pep-0693/
> 
> Or will this be considered as a bug fix?

I'm triggering a buildbot rerun to see if that specific buildbot is actually impacted.

**Comment by sunmy2019** [comment]
https://buildbot.python.org/all/#builders/178/builds/1189

test_monitoring failure is tracked here
https://github.com/python/cpython/issues/103845
test_import failure is tracked here
https://github.com/python/cpython/issues/103879#issuecomment-1528867646

Seems no more failures

**Comment by Fidget-Spinner** [comment]
Please add news. You can copy the previous PR's news item. I will attribute the commit to both of you.

**Comment by Fidget-Spinner** [comment]
Sorry everyone, I messed up a merge commit.

**Comment by arhadthedev** [comment]
Please, don't worry about 229 random files reported as changed. It's a GitHub after-merge artifact, the final commit 59c27fa5cb95e2d608747a50fc675bbe2fc96beb squash-merged into `main` is fine.

**Comment by wangxiang-hz** [comment]
Sorry guys, I just noticed this pull request. I was buy with my work these weeks and just finished a major update at the end of April..
I think @sunmy2019  is right and his commits is better. `call_attribute` should only be called when attribute is not found, then we call it, and clear the `PyExc_AttributeError` being raised.
I made a mistake, `PyObject_GenericGetAttr` may not raise attribute error but other scripts may still raise other exceptions. When this happen, `call_attribute` should not be called.
Thank you all for finding this and sorry for the bugs triggered.

**Pull request merged**